### PR TITLE
Improve automatic application email reliability

### PR DESCRIPTION
**Pull request by clareconstantine**
[PLC-171](https://codedotorg.atlassian.net/browse/PLC-171)
- the cron job to send the scheduled application emails will continue sending emails even if it hits an error, then raise an exception at the end with the error messages of the failed emails (which will be logged to HoneyBadger)
- facilitator application emails use the provided alternate address if the applicant doesn't have an email on their account
- facilitator application emails use the `formatted_applicant_email` method that teacher applications have been using
- add/move/clean up tests

### CODE REVIEW

File: dashboard/app/models/pd/application/teacher1920_application.rb [hunk 1]
```diff
@@ -157,11 +157,6 @@ def principal_approval_state
       nil
     end
 
-    def formatted_teacher_email
```
Comment (clareconstantine) [inline dashboard/app/models/pd/application/teacher1920_application.rb:160]: I renamed this and moved it to application base so facilitator apps can use it too

File: dashboard/test/models/pd/application/teacher1920_application_test.rb [hunk 2]
```diff
@@ -430,41 +430,6 @@ class Teacher1920ApplicationTest < ActiveSupport::TestCase
       refute Email.exists?(associated_unsent_email.id)
     end
 
-    test 'formatted_partner_contact_email' do
```
Comment (clareconstantine) [inline dashboard/test/models/pd/application/teacher1920_application_test.rb:433]: this test is a duplicate of a test in application_base_test.rb

**Review by islemaster — APPROVED** [review]
:+1: Looks great!

File: dashboard/app/models/pd/application/email.rb [hunk 3]
```diff
@@ -34,7 +34,20 @@ def mark_sent!
...
+      rescue => e
+        errors[email.id] = "#{e.message}, #{e.backtrace.first}"
+      end
+
+      if errors.any?
+        msg = "Error sending emails for applications. Errors:\n"
+        errors.each do |application_id, error|
+          msg << "Application #{application_id}: #{error}"
+        end
+        raise msg
```
Comment (breville) [inline dashboard/app/models/pd/application/email.rb:49]: Does HoneyBadger cope okay with an array of messages?

**Review by breville — APPROVED** [review]

File: dashboard/app/models/pd/application/email.rb (hunk 3 above)
Comment (clareconstantine) [reply dashboard/app/models/pd/application/email.rb:49]: good question - this is actually one long string, not an array, but I added newlines to make it easier to read if there are multiple

**Pull request merged**